[HLSL][RootSignature] Add parsing of floats for StaticSampler

inbelic opened the pull request:
- defines in-memory representaiton of MipLODBias to allow for testing of a float parameter
- defines `handleInt` and `handleFloat` to handle converting a token's `NumSpelling` into a valid float
- plugs this into `parseFloatParam` to fill in the MipLODBias param

The parsing of floats is required to match the behaviour of DXC. This behaviour is outlined as follows:
- if the number is an integer then convert it using `_atoi64`, check for overflow and static_cast this to a float
- if the number is a float then convert it using `strtod`, check for float overflow and static_cast this to a float, this will implicitly also check for double over/underflow and if the string is malformed then it will return an error

This pr matches this behaviour by parsing as, uint/int accordingly and then casting, or, by using the correct APFloat semantics/rounding mode with `NumericLiteralParser`.

- adds testing of error diagnostics and valid float param values to demonstrate functionality

Part 2 of https://github.com/llvm/llvm-project/issues/126574

inbelic committed:
pre-req: add missing token to Lexer

inbelic committed:
pre-req: add parsing of MipLODBias as an uint

- defines a float data member of StaticSampler that will be used to test
functionality of parsing a float

inbelic committed:
add support for parsing signed integer for float param

inbelic committed:
add support for parsing a float for float param

inbelic committed:
add testing for over/underflow and addresses other opStatus

- updates the error message to account for floats and fixes the previous
use cases

inbelic committed:
self-review: treat postive integer as unsigned

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -663,6 +669,23 @@ RootSignatureParser::parseStaticSamplerParams() {
+      auto MipLODBias = parseFloatParam();
+      if (!MipLODBias.has_value())
+        return std::nullopt;
+      Params.MipLODBias = (float)*MipLODBias;
```
```suggestion
      Params.MipLODBias = MipLODBias;
```
can remove lingering cast from previous version

Icohedron commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -821,22 +873,110 @@ std::optional<uint32_t> RootSignatureParser::handleUIntLiteral() {
+         "NumSpelling can only consist of digits");
+
+  llvm::APSInt Val = llvm::APSInt(32, true);
+  if (Literal.GetIntegerValue(Val) || INT32_MAX < Val.getExtValue()) {
```
nvm. I see why

Icohedron replied:
But this code wouldn't allow INT32_MIN to be parsed because `abs(INT32_MIN) > INT32_MAX`

inbelic committed:
review: remove INT32_MAX constraint

inbelic committed:
self-review: remove lingering cast

inbelic committed:
self-review: use proper float comparison

inbelic committed:
review: add constraint on magnitude of signed integer

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -821,22 +873,115 @@ std::optional<uint32_t> RootSignatureParser::handleUIntLiteral() {
+         "NumSpelling consists only of [0-9.ef+-]. Any malformed NumSpelling "
+         "will be caught and reported by NumericLiteralParser.");
+
+  // DXC used `strtod` to convert the token string to a float which corresponds
+  // to:
+  auto DXCSemantics = llvm::APFloat::Semantics::S_IEEEdouble;
+  auto DXCRoundingMode = llvm::RoundingMode::NearestTiesToEven;
```
We should probably comment more explicitly (on the function definition?) that we parse the values as double and then convert to float for compatibility with DXC. This function sounds like a generic float parser, so that behaviour could be surprising.

Icohedron commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -821,22 +873,115 @@ std::optional<uint32_t> RootSignatureParser::handleUIntLiteral() {
+  // doesn't overflow as a signed 32-bit int.
+  int64_t MaxMagnitude =
+      -static_cast<int64_t>(std::numeric_limits<int32_t>::min());
+  if (Literal.GetIntegerValue(Val) || MaxMagnitude < Val.getExtValue()) {
```
`2147483648` would be parsed by `handleUIntLiteral` instead. But yes, if `handleUIntLiteral` was called instead for some reason then it would fail to report an overflow.

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -711,6 +734,35 @@ std::optional<Register> RootSignatureParser::parseRegister() {
+    if (!UInt.has_value())
+      return std::nullopt;
+    return (float)UInt.value();
+  } else if (tryConsumeExpectedToken(TokenKind::int_literal)) {
```
Flyby style nit: Don't use `else` after a `return`

https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

inbelic committed:
review: remove use of float.h for numeric_limits

inbelic committed:
review: spell out types

inbelic committed:
review: don't use else after return

inbelic committed:
self-review: fix typo

inbelic committed:
review: use c++ style casting

inbelic committed:
review: initialize directly

inbelic committed:
review: add explicit comment of how the float is handled

inbelic committed:
review: let `handleIntLiteral` properly handle a signed positive integer

- this code path will currently not ever get evaluated but for the sake
of expected behaviour of a future user we will add this functionality
